From 81a829d59e563d1642b0c3c809dacddfde18954c Mon Sep 17 00:00:00 2001
From: Volker Hilsheimer <volker.hilsheimer@qt.io>
Date: Mon, 23 Feb 2026 16:18:48 +0100
Subject: Control styles: call updateTheme from the main thread

While type registration for a plugin is synchronized, we should still
not access GUI resources and singletons from a worker thread (like QML's
type loader thread, in this case).

We have to create the QQuickTheme object during type registration
so that we know what exact type we ended up with, and to be able to read
values from qtquickcontrols2.conf. And we have to initialize the theme
so that control objects use the right defaults for palettes and fonts.

This is a smell, as initializeTheme typically reads from Qt GUI globals,
such as platform theme or QStyleHints. It doesn't seem to be the reason
for data races though, as type loading is a synchronous process as far
as the GUI event loop is concerned, so there are no concurrent writes.

The real issue is that we later call updateTheme() from the thread that
owns the QQuickStylePlugin instance, which is not the main thread. Since
the Observer as the object triggering the call through a signal lives in
the main thread, use a DirectConnection to ensure that updateTheme is
also executed by that thread.

Since initializeTheme is guaranteed to be finished before any call to
updateTheme can happen, there is no concurrency between those two
functions.

Pick-to: 6.11 6.11.0 6.10 6.8
Task-number: QTBUG-143978
Task-number: QTBUG-144397
Change-Id: I35e2b494687259b2788b2ade9d298223d88136fb
Reviewed-by: Doris Verria <doris.verria@qt.io>
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
(cherry picked from commit f6a54c995f04c62de46bfd96b69bf25124e8867b)
---
 src/quickcontrols/qquickstyleplugin.cpp | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git x/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp y/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
index 372d5f39a101df8215eb911a69d49e0756b8f571..0a5343ead59eac516421abde48a0604ba8a50a94 100644
--- x/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
+++ y/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
@@ -104,14 +104,25 @@ void QQuickStylePlugin::registerTypes(const char *uri)
         qCDebug(lcStylePlugin).nospace() << "this style plugin belongs to the fallback style "
             << QQuickStylePrivate::fallbackStyle() << "; calling initializeTheme()";
     }
+
+    if (!isPrimaryFallback && !styleName.isEmpty())
+        QFileSelectorPrivate::addStatics(QStringList() << styleName);
+
+    // Create and initialize the theme, and hook it up to palette and theme changes.
+    // We are currently in the QML loader thread, but we have to return with an
+    // initialized theme so that controls get the right default values for
+    // palette and font. So we have to initialize the theme right now, even though
+    // that will access QGuiApplication globals (such as a the QPlatformTheme or
+    // QStyleHints singletons) from the wrong thread. It is ok, as the loader
+    // process is blocking the GUI thread anyway, so there are no concurrent
+    // writes to those singletons.
     initializeTheme(theme);
 
+    // Updates must however go through the main thread. The Observer moves itself
+    // to the same thread as the QStyleHints instance, which is the main thread.
     themeChangeObserver.reset(new QQuickThemeChangeObserver);
     connect(themeChangeObserver.get(), &QQuickThemeChangeObserver::paletteOrThemeChanged,
-            this, &QQuickStylePlugin::updateTheme);
-
-    if (!isPrimaryFallback && !styleName.isEmpty())
-        QFileSelectorPrivate::addStatics(QStringList() << styleName);
+            this, &QQuickStylePlugin::updateTheme, Qt::DirectConnection);
 }
 
 void QQuickStylePlugin::unregisterTypes()
